src/chainstate/stacks/boot/pox-3.clar

```diff
@@ -0,0 +1,1308 @@
+         (unlock-in-cycle (burn-height-to-reward-cycle unlock-height))
+         (cur-cycle (current-pox-reward-cycle))
+         (first-increased-cycle (+ cur-cycle u1))
+         (stacker-state (unwrap! (map-get? stacking-state
+                                          { stacker: tx-sender })
+                                          (err ERR_STACK_INCREASE_NOT_LOCKED))))
+      ;; tx-sender must be currently locked
+      (asserts! (> amount-stacked u0)
+                (err ERR_STACK_INCREASE_NOT_LOCKED))
+      ;; must be called with positive `increase-by`
+      (asserts! (>= increase-by u1)
+                (err ERR_STACKING_INVALID_AMOUNT))
+      ;; stacker must have enough stx to lock
+      (asserts! (>= amount-unlocked increase-by)
+                (err ERR_STACKING_INSUFFICIENT_FUNDS))
+      ;; stacker must not be delegating
+      (asserts! (is-none (get delegated-to stacker-state))
+                (err ERR_STACKING_IS_DELEGATED))
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+                (err ERR_STACKING_PERMISSION_DENIED))
+      ;; stacker must be directly stacking
+      (asserts! (> (len (get reward-set-indexes stacker-state)) u0)
+                (err ERR_STACKING_ALREADY_DELEGATED))
+      ;; update reward cycle amounts
+      (asserts! (is-some (fold increase-reward-cycle-entry
+            (get reward-set-indexes stacker-state)
+            (some { first-cycle: first-increased-cycle,
+                    reward-cycle: (get first-reward-cycle stacker-state),
+                    stacker: tx-sender,
+                    add-amount: increase-by })))
+            (err ERR_STACKING_UNREACHABLE))
+      ;; NOTE: stacking-state map is unchanged: it does not track amount-stacked in PoX-3
+      (ok { stacker: tx-sender, total-locked: (+ amount-stacked increase-by)})))
+
+;; Extend an active Stacking lock.
+;; *New in Stacks 2.1*
+;; This method extends the `tx-sender`'s current lockup for an additional `extend-count`
+;;    and associates `pox-addr` with the rewards
+(define-public (stack-extend (extend-count uint)
+                             (pox-addr { version: (buff 1), hashbytes: (buff 32) }))
+   (let ((stacker-info (stx-account tx-sender))
+         ;; to extend, there must already be an etry in the stacking-state
+         (stacker-state (unwrap! (get-stacker-info tx-sender) (err ERR_STACK_EXTEND_NOT_LOCKED)))
+         (amount-ustx (get locked stacker-info))
+         (unlock-height (get unlock-height stacker-info))
+         (cur-cycle (current-pox-reward-cycle))
+         (unlock-in-cycle (burn-height-to-reward-cycle unlock-height))
+         ;; if the account unlocks *during* this cycle (should only occur during testing),
+         ;; set first-extend-cycle to the next cycle.
+         (first-extend-cycle (if (> (+ cur-cycle u1) unlock-in-cycle)
+                                    (+ cur-cycle u1) unlock-in-cycle))
```

Comment:
I think this is something we may want to push a little bit on -- in `pox-3`, can we assume the `first-extend-cycle` is the next cycle? (because unlocks all occurred at epoch 2.2) If so, I think we should do this check, but yield an `err` on failing the check.

Reply:
> If so, I think we should do this check, but yield an err on failing the check.

Yes, I think that makes sense.

Reply:
Looks that check already exists below:

```clarity
(asserts! (> first-extend-cycle cur-cycle) (err ERR_STACKING_INVALID_LOCK_PERIOD))
```

Reply:
Yep, so line 1027 would become just:

```clarity
(first-extend-cycle unlock-in-cycle)
```

Reply:
I had deleted the variable `first-extend-cycle` and updated the uses. Any problem with that?

Reply:
Maybe better to rename `unlock-in-cycle` to `first-extend-cycle`.

Reply:
> Maybe better to rename unlock-in-cycle to first-extend-cycle.

Yeah, that seems preferable, with a comment like `;; first-extend-cycle will be the cycle in which tx-sender *would have* unlocked`